src/main/java/net/neoforged/neoforge/capabilities/CapabilityHooks.java

```diff
@@ -0,0 +1,170 @@
+                return new InvWrapper(container);
+            });
+        }
+
+        // Entities
+        var containerEntities = List.of(
+                EntityType.CHEST_BOAT,
+                EntityType.CHEST_MINECART,
+                EntityType.HOPPER_MINECART);
+        for (var entityType : containerEntities) {
+            event.registerEntity(Capabilities.ItemHandler.ENTITY, entityType, (entity, ctx) -> new InvWrapper(entity));
+            event.registerEntity(Capabilities.ItemHandler.ENTITY_AUTOMATION, entityType, (entity, ctx) -> new InvWrapper(entity));
+        }
+        event.registerEntity(Capabilities.ItemHandler.ENTITY, EntityType.PLAYER, (player, ctx) -> new PlayerInvWrapper(player.getInventory()));
+        // Register to all entity types to make sure we support all living entity subclasses.
+        for (EntityType<?> entityType : BuiltInRegistries.ENTITY_TYPE) {
+            event.registerEntity(Capabilities.ItemHandler.ENTITY, entityType, (entity, ctx) -> {
+                if (entity instanceof AbstractHorse horse)
+                    return new InvWrapper(horse.getInventory());
+                else if (entity instanceof LivingEntity livingEntity)
+                    return new CombinedInvWrapper(new EntityHandsInvWrapper(livingEntity), new EntityArmorInvWrapper(livingEntity));
+
+                return null;
+            });
+        }
+
+        // Items
+        for (Item item : BuiltInRegistries.ITEM) {
+            if (item.getClass() == BucketItem.class)
+                event.registerItem(Capabilities.FluidHandler.ITEM, (stack, ctx) -> new FluidBucketWrapper(stack), item);
+        }
+        if (NeoForgeMod.MILK.isPresent()) {
+            event.registerItem(Capabilities.FluidHandler.ITEM, (stack, ctx) -> new FluidBucketWrapper(stack), Items.MILK_BUCKET);
+        }
+        event.registerItem(Capabilities.ItemHandler.ITEM, (stack, ctx) -> new ShulkerItemStackInvWrapper(stack),
+                Items.SHULKER_BOX,
+                Items.BLACK_SHULKER_BOX,
+                Items.BLUE_SHULKER_BOX,
+                Items.BROWN_SHULKER_BOX,
+                Items.CYAN_SHULKER_BOX,
+                Items.GRAY_SHULKER_BOX,
+                Items.GREEN_SHULKER_BOX,
+                Items.LIGHT_BLUE_SHULKER_BOX,
+                Items.LIGHT_GRAY_SHULKER_BOX,
+                Items.LIME_SHULKER_BOX,
+                Items.MAGENTA_SHULKER_BOX,
+                Items.ORANGE_SHULKER_BOX,
+                Items.PINK_SHULKER_BOX,
+                Items.PURPLE_SHULKER_BOX,
+                Items.RED_SHULKER_BOX,
+                Items.WHITE_SHULKER_BOX,
+                Items.YELLOW_SHULKER_BOX);
+    }
+
+    public static void invalidateCapsOnChunkLoad(ChunkEvent.Load event) {
+        if (!event.getLevel().isClientSide()) {
+            ((ServerLevel) event.getLevel()).invalidateCapabilities(event.getChunk().getPos());
+        }
+    }
+
+    public static void invalidateCapsOnChunkUnload(ChunkEvent.Unload event) {
```

Comment:
I would say that it is the responsibility of the caller to make sure that capability queries don't go into unloaded chunks? It means that mods who perform capability queries from something else than a block entity tick need to implement their own checks for this. I would say this is reasonable, since they probably already maintain their own cache of which positions are ticking. (e.g. a pipe network probably knows which parts of the network are ticking)

Reply:
I think I likely agree with this, just wanted to bring it up, as it could catch people as off guard if their cached capabilities are still valid, but the chunk of the block the capability is in doesn't show as loaded. Granted you might be right that for cases when handled in block entity tick even on the border, then the other chunk is still accessible as it is a few chunk radius before it gets into loaded but inaccessible territory (at which point block entities aren't ticking on the border).

And actually even if not... in general if you mutate a capability then it will still change the save state, and as the chunks aren't fully unloaded yet and are just inaccessible, I think that means they would still be able to be get marked for saving... Marking this as resolved as I don't think it is actually prevalent to this PR.